File: v2/backend/app/routers/tokens.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from ..auth import require_user
from ..token_store import create_token, list_tokens, revoke_token, set_token_trusted

router = APIRouter(prefix="/tokens", tags=["tokens"])


class TokenCreate(BaseModel):
    name: str


@router.get("")
def get_tokens(user: dict = Depends(require_user)):
    return list_tokens(user["sub"])


@router.post("", status_code=201)
def new_token(body: TokenCreate, user: dict = Depends(require_user)):
    name = body.name.strip() or "API Token"
    raw, record = create_token(user["sub"], name)
    # raw token returned exactly once - never stored in plaintext, never returned again
    return {**record, "token": raw}


@router.delete("/{token_id}", status_code=204)
def delete_token(token_id: str, user: dict = Depends(require_user)):
    if not revoke_token(user["sub"], token_id):
        raise HTTPException(404, "Token not found")


class TokenTrustUpdate(BaseModel):
    trusted: bool


@router.patch("/{token_id}/trusted", status_code=200)
def update_token_trusted(token_id: str, body: TokenTrustUpdate, user: dict = Depends(require_user)):
    if user.get("auth_method") != "cookie":
        raise HTTPException(403, "Trusted flag can only be changed from the browser UI")
    if not set_token_trusted(user["sub"], token_id, body.trusted):
        raise HTTPException(404, "Token not found")
    return {"ok": True, "trusted": body.trusted}

File: v2/frontend/src/lib/tokens.ts
const BASE = '/api/v1/tokens'

export interface ApiToken {
  token_id:     string
  name:         string
  prefix:       string
  created_at:   string
  last_used_at: string | null
  trusted:      boolean
}

export interface NewTokenResponse extends ApiToken {
  token: string  // plaintext - shown once only
}

export async function fetchTokens(): Promise<ApiToken[]> {
  const res = await fetch(BASE, { credentials: 'include' })
  if (!res.ok) throw new Error('Failed to load tokens')
  return res.json()
}

export async function createToken(name: string): Promise<NewTokenResponse> {
  const res = await fetch(BASE, {
    method: 'POST',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) throw new Error('Failed to create token')
  return res.json()
}

export async function revokeToken(token_id: string): Promise<void> {
  const res = await fetch(`${BASE}/${token_id}`, { method: 'DELETE', credentials: 'include' })
  if (!res.ok && res.status !== 404) throw new Error('Failed to revoke token')
}

export async function setTokenTrusted(token_id: string, trusted: boolean): Promise<void> {
  const res = await fetch(`${BASE}/${token_id}/trusted`, {
    method: 'PATCH',
    credentials: 'include',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ trusted }),
  })
  if (!res.ok) throw new Error('Failed to update token trust')
}
